### PR TITLE
publications app

### LINKED ISSUE #612: feat: Publications app
**Describe the solution you'd like**

New app for viewing magazines/other publications such as UpDaTe 

**Additional context**

There is no good pdf viewer available that fits our licensing requirements, so we need to create something using pdf.js etc.

### PR DESCRIPTION
**Pull request by Theoabw**

### CODE REVIEW

**Comment by theoabw** [comment]
TODO: 
1. Styling on mobile and overall
2. Custom slug functionality
3. Make sure it works with S3

File: static/common/publications/pdf.mjs
Comment (jaspnas) [inline static/common/publications/pdf.mjs:1]: Why are we not downloading these from a CDN?

File: static/common/publications/pdf.worker.mjs
Comment (jaspnas) [inline static/common/publications/pdf.worker.mjs:1]: Why are we not downloading these from a CDN?

**Review by jaspnas — CHANGES_REQUESTED** [review]

File: publications/views.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+from .models import PDFFile
+
+
+@login_required
+def pdf_view(request, pk):
+    pdf_file = get_object_or_404(PDFFile, pk=pk)
+    try:
+        pdf_url = pdf_file.get_file_url()
+    except ClientError as e:
+        return JsonResponse({'error': str(e)}, status=404)
```
[the remark this answers is not in the document]
Comment (jaspnas) [reply publications/views.py]: Let's not return JsonResponse, but let django handle it to show the default 404 page

File: publications/models.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+from .fields import PrivateFileField
+
+
+def upload_to(instance, filename):
+    return f'pdfs/{filename}'
+
+
+class PDFFile(models.Model):
+    title = models.CharField(_('Title'), max_length=250)
+    file = PrivateFileField(upload_to=upload_to)
```
Comment (jaspnas) [inline publications/models.py]: We probably don't want to store these files in private storage as publications are intended to be publicly viewable.

File: publications/models.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+from django.utils.translation import gettext_lazy as _
+from django.conf import settings
+from .fields import PrivateFileField
+
+
+def upload_to(instance, filename):
+    return f'pdfs/{filename}'
+
+
+class PDFFile(models.Model):
```
Comment (jaspnas) [inline publications/models.py:16]: We want to add slug support

**Review by jaspnas — CHANGES_REQUESTED** [review]

File: core/settings/common.py [hunk 4]
```diff
@@ -233,6 +233,7 @@ def get_staff_groups(default_groups: list):
 else:
     MEDIA_ROOT = os.path.join(BASE_DIR, 'media')
     MEDIA_URL = '/media/'
+    DEFAULT_FILE_STORAGE = 'django.core.files.storage.FileSystemStorage'
```
Comment (jaspnas) [inline core/settings/common.py]: This is deprecated and removed in django 5.1

File: publications/models.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import pymupdf
+
+def upload_to(instance, filename):
+    return f'pdfs/{instance.slug}/{filename}'
+
+class PDFFile(models.Model):
+    title = models.CharField(_('Title'), max_length=250)
+    slug = models.SlugField(_('Slug'), max_length=255, unique=True, blank=True,
+                            help_text=_('Leave empty to auto-generate from title'))
+    file = models.FileField(_('File'), upload_to=upload_to, storage=PublicMediaStorage() if settings.USE_S3 else None)
```
Comment (jaspnas) [inline publications/models.py]: We shouldn't have if statements here, see how archive is doing this by redefining save()

File: publications/views.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
+from django.shortcuts import render, get_object_or_404
+from django.contrib.auth.decorators import login_required
+from django.http import Http404, JsonResponse
+from .models import PDFFile
+from django.core.paginator import Paginator
+
+@login_required
```
Comment (jaspnas) [inline publications/views.py]: I don't know if we want login required for all PDF's, I think UpDate is currently stored in a public static page

File: publications/models.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            self.slug = self.generate_unique_slug()
+
+        if self.file:
+            self.file_size = self.file.size
+            self.num_pages = self.count_pdf_pages()
+
+
+        super().save(*args, **kwargs)
+
+    def count_pdf_pages(self):
```
Comment (jaspnas) [inline publications/models.py]: Do we need to count the pages on the backend, or can we have the client do it?

File: publications/views.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    }
+    return render(request, 'publications/viewer.html', context)
+
+@login_required
+def pdf_list(request):
+    pdfs = PDFFile.objects.all()
+    paginator = Paginator(pdfs, 10)  # Show 10 PDFs per page
+    page_number = request.GET.get('page')
+    page_obj = paginator.get_page(page_number)
+    return render(request, 'publications/list.html', {'page_obj': page_obj})
```
Comment (jaspnas) [inline publications/views.py]: Template not found

**Review by jaspnas — CHANGES_REQUESTED** [review]

File: publications/models.py (hunk 7 above)
Comment (theoabw) [reply publications/models.py]: We don't actually need this feature and it's not really critical for it to work so I'll probably just remove it for now. Future issue can add a client side counter if anyone wants to do that.

File: core/settings/common.py [hunk 9]
```diff
@@ -229,10 +229,35 @@ def get_staff_groups(default_groups: list):
...
+        },
+        "staticfiles": {
+            "BACKEND": "django.contrib.staticfiles.storage.StaticFilesStorage",
+        },
+    }
+
 else:
     MEDIA_ROOT = os.path.join(BASE_DIR, 'media')
     MEDIA_URL = '/media/'
+    DEFAULT_FILE_STORAGE = 'django.core.files.storage.FileSystemStorage'
```
Comment (jaspnas) [inline core/settings/common.py]: Should probably remove this

File: requirements.txt [hunk 10]
```diff
@@ -26,3 +26,4 @@ django_cleanup
 daphne
 celery[redis]
 gunicorn
+pymupdf
```
Comment (jaspnas) [inline requirements.txt]: This should be removed if we don't need it

File: publications/models.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    slug = models.SlugField(_('Slug'), max_length=255, unique=True, blank=True,
+                            help_text=_('Leave empty to auto-generate from title'))
+    file = PublicFileField(_('File'), upload_to=upload_to)
+    description = models.TextField(_('Description'), blank=True)
+    uploaded_at = models.DateTimeField(_('Uploaded at'), auto_now_add=True)
+    updated_at = models.DateTimeField(_('Updated at'), auto_now=True)
+    author = models.CharField(_('Author'), max_length=250, blank=True)
+    publication_date = models.DateField(_('Publication Date'), null=True, blank=True)
+    file_size = models.PositiveIntegerField(_('File Size (bytes)'), editable=False, null=True)
+    is_public = models.BooleanField(_('Public Access'), default=False,
```
Comment (jaspnas) [inline publications/models.py]: This should be true by default

File: publications/models.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    title = models.CharField(_('Title'), max_length=250)
+    slug = models.SlugField(_('Slug'), max_length=255, unique=True, blank=True,
+                            help_text=_('Leave empty to auto-generate from title'))
+    file = PublicFileField(_('File'), upload_to=upload_to)
+    description = models.TextField(_('Description'), blank=True)
+    uploaded_at = models.DateTimeField(_('Uploaded at'), auto_now_add=True)
+    updated_at = models.DateTimeField(_('Updated at'), auto_now=True)
+    author = models.CharField(_('Author'), max_length=250, blank=True)
+    publication_date = models.DateField(_('Publication Date'), null=True, blank=True)
+    file_size = models.PositiveIntegerField(_('File Size (bytes)'), editable=False, null=True)
```
Comment (jaspnas) [inline publications/models.py]: Not needed

File: publications/models.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+class PDFFile(models.Model):
+    title = models.CharField(_('Title'), max_length=250)
+    slug = models.SlugField(_('Slug'), max_length=255, unique=True, blank=True,
+                            help_text=_('Leave empty to auto-generate from title'))
+    file = PublicFileField(_('File'), upload_to=upload_to)
+    description = models.TextField(_('Description'), blank=True)
+    uploaded_at = models.DateTimeField(_('Uploaded at'), auto_now_add=True)
+    updated_at = models.DateTimeField(_('Updated at'), auto_now=True)
+    author = models.CharField(_('Author'), max_length=250, blank=True)
```
Comment (jaspnas) [inline publications/models.py]: Is this needed?

File: publications/models.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    publication_date = models.DateField(_('Publication Date'), null=True, blank=True)
```
Comment (jaspnas) [inline publications/models.py]: Should be autofilled as now

**Comment by jaspnas** [comment]
Color theme doesn't look quite right 
image

File: publications/admin.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    list_filter = ('is_public', 'requires_login', 'uploaded_at', 'updated_at', 'publication_date')
+    search_fields = ('title', 'slug', 'description', 'author')
+    prepopulated_fields = {'slug': ('title',)}
+    readonly_fields = ('uploaded_at', 'updated_at', 'file_size')
+    fieldsets = (
+        (None, {
+            'fields': ('title', 'slug', 'author', 'publication_date', 'description', 'file')
+        }),
+        ('Access Control', {
+            'fields': ('is_public', 'requires_login'),
```
Comment (jaspnas) [inline publications/admin.py:16]: These feel like duplicates

File: static/common/publications/js/pdfViewer.js [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    } catch (error) {
+        console.error('Error loading PDF:', error);
+        showErrorMessage('Failed to load PDF. Please try again later.');
+        state.updateState({ isLoading: false });
+    }
+}
+
+function updateLoadingProgress(percent) {
+    // Implement loading progress UI update
+    console.log(`Loading PDF: ${percent.toFixed(2)}%`);
```
Comment (jaspnas) [inline static/common/publications/js/pdfViewer.js]: This is spammy in the console

**Review by jaspnas — CHANGES_REQUESTED** [review]
Seems to work on s3, shouldn't break anything existing. Let's make admin a bit clearer and then we can merge this.

File: publications/admin.py (hunk 15 above)
Comment (theoabw) [reply publications/admin.py:16]: One is if anyone (including members) gets to see it and the other is for if you have to be logged in to see it

**Comment by theoabw** [comment]
> Color theme doesn't look quite right image

dunno what the colors should be. Feel free to change it if you want to

**Comment by andreedd** [comment]
You could bring it to life a bit more with some icons.

It could kind of look like the events:
Screenshot 2024-08-23 at 17 27 09

Colors can be find in root.css.

File: core/settings/common.py [hunk 17]
```diff
@@ -249,14 +258,12 @@ def get_staff_groups(default_groups: list):
                 "location": PUBLIC_MEDIA_LOCATION,
             }
         },
-        "staticfiles": {
```
Comment (jaspnas) [inline core/settings/common.py:252]: You can't partially redefine storages here if you define it on root level

**Review by jaspnas — APPROVED** [review]

**Pull request merged**